api/src/modules/analysis/providers/planning-units/update-planning-units.service.ts

```diff
@@ -0,0 +1,42 @@
+import { Injectable } from '@nestjs/common';
+import { AdjustPlanningUnits } from '../../entry-points/adjust-planning-units';
+import { AdjustPlanningUnitsInput } from '../../entry-points/adjust-planning-units-input';
+
+import { ArePuidsAllowedPort } from './are-puids-allowed.port';
+import { RequestJobPort } from './request-job.port';
+
+type Success = true;
+
+@Injectable()
+export class UpdatePlanningUnitsService implements AdjustPlanningUnits {
+  constructor(
+    private readonly puUuidValidator: ArePuidsAllowedPort,
+    private readonly jobRequester: RequestJobPort,
+  ) {}
+
+  async update(
+    scenarioId: string,
+    constraints: AdjustPlanningUnitsInput,
+  ): Promise<Success> {
+    const targetPuIds = [
+      ...(constraints.include?.pu ?? []),
+      ...(constraints.exclude?.pu ?? []),
+    ];
+    if (targetPuIds.length > 0) {
+      const { errors } = await this.puUuidValidator.validate(
+        scenarioId,
+        targetPuIds,
+      );
+      if (errors.length > 0) {
+        throw new Error('Given PU ids are not reachable in this context.');
```

Comment:
Should we treat this as a `BadRequestException`?

Also, I suggest to make the error message a bit clearer, for example

```suggestion
        throw new Error('One or more of the planning units provided for exclusion or inclusion does not match any planning unit of the present scenario.');
```